cloudinit/netinfo.py

```diff
@@ -18,13 +20,74 @@
 
 LOG = logging.getLogger()
 
-
+# Example netdev format:
+# {'eth0': {'hwaddr': '00:16:3e:16:db:54',
+#           'ipv4': [{'bcast': '10.85.130.255',
+#                     'ip': '10.85.130.116',
+#                     'mask': '255.255.255.0',
+#                     'scope': 'global'}],
+#           'ipv6': [{'ip': 'fd42:baa2:3dd:17a:216:3eff:fe16:db54/64',
+#                     'scope6': 'global'},
+#                    {'ip': 'fe80::216:3eff:fe16:db54/64', 'scope6': 'link'}],
+#           'up': True},
+#  'lo': {'hwaddr': '',
+#         'ipv4': [{'bcast': '',
+#                   'ip': '127.0.0.1',
+#                   'mask': '255.0.0.0',
+#                   'scope': 'host'}],
+#         'ipv6': [{'ip': '::1/128', 'scope6': 'host'}],
+#         'up': True}}
 DEFAULT_NETDEV_INFO = {"ipv4": [], "ipv6": [], "hwaddr": "", "up": False}
 
 
+def _netdev_info_iproute_json(ipaddr_json):
+    """Get network device dicts from ip route and ip link info.
+
+    ipaddr_out: Output string from 'ip --json addr' command.
+
+    Returns a dict of device info keyed by network device name containing
+    device configuration values.
+
+    Raises json.JSONDecodeError if json could not be decoded
+    """
+    ipaddr_data = json.loads(ipaddr_json)
+    devs = {}
+
+    for dev in ipaddr_data:
+        flags = dev["flags"]
+        dev_info = {
+            "hwaddr": dev["address"] if dev["link_type"] == "ether" else "",
+            "up": bool("UP" in flags and "LOWER_UP" in flags),
+            "ipv4": [],
+            "ipv6": [],
+        }
+        for addr in dev["addr_info"]:
+            if addr["family"] == "inet":
```

Comment:
Generally the logic here looks good and mirrors what we have in `_netdev_info_iproute`. I can also note that on my laptop your JSON output handling actually fixes an existing bug in properly parsing  inet6 lines from `ip addr show`.


We have a gap though and I think it's worth adding a unittest that validatates that comparable output from `ip --json addr show` and `ip addr show` gets the same resulting dict from ` _netdev_info_iproute_json` and `_netdev_info_iproute` respectively.

When I try running the following asserts on my laptop given both `ip addr show` and `ip --json addr show` output I can see a failure in `_netdev_info_iproute to parse inet6 lines leaving that value empty.
```python
from cloudinit.netinfo import _netdev_info_iproute_json, _netdev_info_iproute   
json_dict = _netdev_info_iproute_json(ip_show_json)                             
orig_dict = _netdev_info_iproute(ip_show_out)                                   
                                                                                
assert json_dict.keys() == orig_dict.keys()                                     
for k in orig_dict.keys():                                                      
  if orig_dict[k] != json_dict[k]:                                              
      assert orig_dict[k].keys() == json_dict[k].keys()                         
      for subk in orig_dict[k].keys():                                          
          assert orig_dict[k][subk] == json_dict[k][subk], f"{k}.{subk} not equal"

$ python3 ./test.py
AssertionError: tun0.ipv6 not equal
```

Unrelated to this PR, but the failures I see locally on Focal are due to this line not matching our regex
`'    inet6 2001:67c:1562:8007::aac:41aa peer 2001:67c:1562:8007::1/64 scope global noprefixroute '`
```python
m = re.match(
                r"\s+inet6\s(?P<ip>\S+)\sscope\s(?P<scope6>\S+).*", line
)
```

I think we may need to fix our regex to also handle optional "(peer\s(?P<peer_ip>\S+)?". Not sure if you want to tackle that in this PR or not.

Reply:
Works on my machine! Ship it!

But in all seriousness, do we want to maintain both implementations? The whole point of the previous PR was to fix an issue where I was seeing the `metric` field on the inet line. It's not actually a new field, just jammy defaults to using it. So should we be fixing that in the regex as well?

Reply:
@TheRealFalcon agreed we don't want to maintain both implementations. This is seen as an existing bug on distros with older `iproute2`. I don't think it's worth the regex fix the older version parsing. So yeah, let's forget about that because 99% of our distributiions which carry an `ip` tool should support the --json flag.

Reply:
Heh...I recently pushed the commit to fix it. I think at this point it's easier to just keep it and not update it anymore going forward.